### PR TITLE
Notebooks: Add smoke test

### PR DESCRIPTION
**Pull request by lucyzhang929**
This PR adds a smoke test that opens up a notebook file (https://github.com/kburtram/azuredatastudio-smoke-test-repo/blob/master/hello.ipynb), runs all cells, saves notebooks, opens notebook and checks that all outputs are saved.

### CODE REVIEW

File: test/smoke/src/sql/areas/notebook/notebook.test.ts [hunk 1]
```diff
@@ -22,7 +21,25 @@ export function setup() {
...
+			await app.workbench.sqlNotebook.runAllCells();
+			await app.workbench.sqlNotebook.waitForAllResults([1, 2, 3]);
+
+			await app.workbench.quickaccess.runCommand('workbench.action.files.save');
 			await app.workbench.quickaccess.runCommand('workbench.action.closeActiveEditor');
+
+			await app.workbench.sqlNotebook.openFile('hello.ipynb');
+			await app.workbench.sqlNotebook.waitForKernel('Python 3');
+			// The cell ids increase by one after closing and reopening notebook
+			await app.workbench.sqlNotebook.waitForAllResults([4, 5, 6]);
```
Comment (barbaravaldez) [inline test/smoke/src/sql/areas/notebook/notebook.test.ts:42]: Looks great! I'm just curious about this line, do cell ids always increase by one after closing and reopening?

Comment (lucyzhang929) [reply test/smoke/src/sql/areas/notebook/notebook.test.ts:42]: I think the cell ids are dependent on how many CellModel objects are created after ADS is opened. So the first cell created in the previous test has id = 0, the first time hello.ipynb is opened, 3 CellModels are created with ids 1,2,3, and when hello.ipynb is re-opened, another 3 CellModels are created with ids 4,5,6.

Comment (Charles-Gagnon) [reply test/smoke/src/sql/areas/notebook/notebook.test.ts:42]: Do we start a new instance of ADS for every smoke test ran? if not what happens if we add another smoke test that opens a notebook and add cells - this is going to break at that point then right?

Instead what if you took in an index and then had the function grab the ordered list of cell elements and then used the passed in indexes to pick which ones to wait for?

File: test/automation/src/sql/notebook.ts [hunk 2]
```diff
@@ -78,6 +84,23 @@ export class Notebook {
 		const outputComponent = '.notebook-cell.active .notebook-output';
 		await this.code.waitForElement(outputComponent);
 	}
+
+	async waitForAllResults(cellIds: number[]): Promise<void> {
```
Comment (Charles-Gagnon) [inline test/automation/src/sql/notebook.ts:88]: When I see a function named something like "waitForAll..." I would assume that it does actually wait for ALL the results which isn't the case here.

I'd suggest : 

1. Renaming this to `waitForResults` with the same parameter
1. Adding a new function `waitForAllResults` that does wait for all results without needing to pass in cell IDs

**Review by Charles-Gagnon — CHANGES_REQUESTED** [review]

File: test/automation/src/sql/notebook.ts (hunk 2 above)
Comment (lucyzhang929) [reply test/automation/src/sql/notebook.ts:88]: The same instance of ADS is used for each smoke test. If another notebook smoke test is added before this one then it would break this test because of the cell id changes. But it would not break if tests were added after this one.

The issue I ran into while implementing waitForAllResults is that there is no way to identify each cell so I added the cell.id to select each cell. I can change the method name to waitForResults.

In order to implement a waitForAllResults method that doesn't rely on the cell ids, I think we would need some keyboard shortcuts (ex. focus on first cell, focus next cell).

Comment (Charles-Gagnon) [reply test/automation/src/sql/notebook.ts]: For waiting for all results you should be able to get the list of cells by using the selector `div.notebook-cell` and then for each of those elements wait for the child element with the `notebook-output` class.

Why do you think we need keyboard shortcuts?

Comment (lucyzhang929) [reply test/automation/src/sql/notebook.ts]: Hm I need to do a little more investigation but I remember running into an issue where using div.notebook-cell will only select the first cell. Let me try again.

Comment (Charles-Gagnon) [reply test/automation/src/sql/notebook.ts]: Well it depends on what you're doing. waitForElement will pick the first one. So make sure you're using `waitForElements` https://github.com/microsoft/azuredatastudio/blob/d2bdd2bace223ecc9436eee27b47b2618a353de6/test/automation/src/code.ts#L312

Comment (lucyzhang929) [reply test/automation/src/sql/notebook.ts]: That works! Thank you Charles!

File: test/automation/src/sql/notebook.ts [hunk 3]
```diff
@@ -74,10 +80,49 @@ export class Notebook {
...
+	async waitForResults(cellIds: string[]): Promise<void> {
+		for (let i of cellIds) {
+			await this.code.waitForElement(`div.notebook-cell[id="${i}"] .notebook-output`);
+		}
+	}
+
+	async waitForAllResults(): Promise<void> {
+		let cellIds: string[] = [];
+		await this.code.waitForElements('div.notebook-cell', false, result => {
+			for (let cell of result) {
```
Comment (Charles-Gagnon) [inline test/automation/src/sql/notebook.ts:97]: `cellIds = result.map(cell => cell.attributes['id']);`

File: test/automation/src/sql/notebook.ts [hunk 4]
```diff
@@ -74,10 +80,49 @@ export class Notebook {
...
+	async waitForResultsGone(cellIds: string[]): Promise<void> {
+		for (let i of cellIds) {
+			await this.code.waitForElementGone(`div.notebook-cell[id="${i}"] .notebook-output`);
+		}
+	}
+
+	async waitForAllResultsGone(): Promise<void> {
+		let cellIds: string[] = [];
+		await this.code.waitForElements('div.notebook-cell', false, result => {
+			for (let cell of result) {
```
Comment (Charles-Gagnon) [inline test/automation/src/sql/notebook.ts:119]: `cellIds = result.map(cell => cell.attributes['id']);`

**Review by Charles-Gagnon — APPROVED** [review]
